public/app/plugins/panel/barchart/utils.ts

```diff
@@ -508,26 +463,131 @@ export function prepareBarChartDisplayValues(
+  fields.unshift(xField);
 
   return {
-    aligned: frame,
+    aligned,
     colorByField,
     viz: [
       {
-        length: firstField.values.length,
+        length: xField.values.length,
         fields: fields, // ideally: fields.filter((f) => !Boolean(f.config.custom?.hideFrom?.viz)),
       },
     ],
     legend: {
       fields: legendFields,
-      length: firstField.values.length,
+      length: xField.values.length,
     },
   };
 }
 
+function getBarFields(
+  frame: DataFrame,
+  xField: Field | undefined,
+  options: PanelOptions,
+  theme: GrafanaTheme2
+): {
+  xField: Field | undefined;
+  fields: Field[];
+  aligned: DataFrame;
+} {
+  let stringFields: Field[] | undefined = undefined;
+  let timeField: Field | undefined = undefined;
+  let fields: Field[] = []; // number fields
+  for (const field of frame.fields) {
+    if (field === xField) {
+      continue;
+    }
+
+    switch (field.type) {
+      case FieldType.string:
+        if (!stringFields) {
+          stringFields = [];
+        }
+        stringFields.push(field);
+        break;
+
+      case FieldType.time:
+        if (!timeField) {
+          timeField = field;
+        }
+        break;
+
+      case FieldType.number: {
+        const copy = {
+          ...field,
+          state: {
+            ...field.state,
+            seriesIndex: fields.length, // off by one?
+          },
+          config: {
+            ...field.config,
+            custom: {
+              ...field.config.custom,
+              stacking: {
+                group: '_',
+                mode: options.stacking,
+              },
+            },
+          },
+          values: new ArrayVector(
+            field.values.toArray().map((v) => {
+              if (!(Number.isFinite(v) || v == null)) {
+                return null;
+              }
+              return v;
+            })
+          ),
+        };
+
+        if (options.stacking === StackingMode.Percent) {
+          copy.config.unit = 'percentunit';
+          copy.display = getDisplayProcessor({ field: copy, theme });
+        }
+
+        fields.push(copy);
+      }
+    }
+  }
+
+  if (!xField) {
+    if (stringFields?.length) {
+      xField = stringFields[0];
+
+      // When multiple strings exist display them as labels
+      if (stringFields?.length > 1) {
+        const names = new Array(frame.length);
+        for (let i = 0; i < frame.length; i++) {
+          const labels: Labels = {};
+          for (const f of stringFields) {
+            labels[f.name] = f.values.get(i);
+          }
+          names[i] = formatLabels(labels);
+        }
+        xField = {
+          name: 'Name',
+          type: FieldType.string,
+          config: {},
+          values: new ArrayVector(names),
+        };
+        xField.display = getDisplayProcessor({ field: xField, theme });
+      }
+    } else if (timeField) {
+      xField = timeField;
+    }
+
+    // Try converting to numericWide
```

Comment:
this avoids showing "frame is missing a string field" if the value can be converted to numericLong